### PR TITLE
Experimental Random Crawl task implemented with Reinforcement Learning (DQN)

### LINKED ISSUE #631: Random Crawl
## Feature request
Random Crawl
#### Feature description

Utilizes Reinforcement Learning to make turtlebot 3 to randomly crawl around without hitting any obstacles.

- Have already implemented this task internally. Will need to bring it out to our repo.

- Random Crawl can be used as:
   1. Map exploration
   2. Localization test
   3. System test
   4. Tutorial to enable RL for different tasks

### PR DESCRIPTION
**Pull request by mhpanah**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (fixes #631) |
| Primary OS tested on | (Ubuntu 18.04) |
| Robotic platform tested on | (Turtlebot3 gazebo simulation) |

---

## Description of contribution in a few bullet points
Utilizes Reinforcement Learning to make turtlebot 3 to randomly crawl around without hitting any obstacles.

## Future work that may be required in bullet points

- Improve collision detection method. Currently, the collision detection is based on raw laser data which can be noisy. This noise sometimes causes false positive collision which needs to be further investigated.

- Improve reward function

- Add other RL algorithms

- Will add more documentation regarding the reward function

### CODE REVIEW

**Comment by SteveMacenski** [comment]
@mhpanah I've also been playing around with this type of tech the last few weeks, good to see similarities in both approaches.

I wanted to ask if perhaps this doesn't belong in this repo and start a `navigation2_experimental` repo to place this in for the time being. I wouldn't feel too comfortable with having this included in the navigation stack directly for the following reasons:
- assumes a horizontal laser scanner which isn't generalized across all robots (pretty weak reason, but  adding)
- really only works for the turtlebot3 if using pre-trained weights 
- collision detection based on distance to laser rather than a robust state validity check
- and the main reason: this technology itself is very experimental and I wouldn't want someone to think this was at a product grade level where they could trust this around people in dynamic environments without severe risk.

File: nav2_experimental_rl/random_crawl.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+from rclpy.node import Node
+from time import sleep
+from keras.models import load_model
+
+def loadModel(env):
+    state = env.reset()
+    observation_space = len(state)
+    state = np.reshape(state, [1, observation_space])
+
+    project_path = os.path.abspath(os.path.dirname(__file__))
```
Comment (SteveMacenski) [inline nav2_experimental_rl/random_crawl.py]: I saw this use in the turtlebot demos, can we have this changed to use properly `rospkg` an example

```
rospack = rospkg.RosPack()
gazPath = rospack.get_path('turtlebot3_gazebo')
self.modelPath = gazPath + '/models/turtlebot3_square/goal_box/model.sdf'
```

Comment (SteveMacenski) [reply nav2_experimental_rl/random_crawl.py]: its cleaner and more clear if the install paths are wacky as python tends to get

File: nav2_experimental_rl/random_crawl_train.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import random
+import numpy as np
+import tensorflow as tf
+
+import rclpy
+from rclpy.node import Node
+from time import sleep
+import parameters
+
+def trainModel(env, action_size):
```
Comment (SteveMacenski) [inline nav2_experimental_rl/random_crawl_train.py]: for the whiles can we add the `rclpy.ok()` condition for clean exits?

File: nav2_experimental_rl/random_crawl_train.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    for _ in range(parameters.EPISODES):
+        print("Episode number: " + str(_))
+        state = env.reset()
+        observation_size = len(state)
+        state = np.reshape(state, [1, observation_size])
+        done = False
+        while not done:
+            agent.step += 1
+            target_model_update_counter += 1
+            if target_model_update_counter%parameters.TARGET_MODEL_UPDATE_STEP == 0:
```
Comment (SteveMacenski) [inline nav2_experimental_rl/random_crawl_train.py]: We're switching from 4 spaces a line to 2 in the inner if conditions in a few of these files.

File: nav2_experimental_rl/turtlebot3_env.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        self.states_input = [3.5] * 8
+        self.zero_div_tol = 0.01
+        self.range_min = 0.0
+
+        self.pub_cmd_vel = node_.create_publisher(Twist, 'cmd_vel')
+        self.sub_scan = node_.create_subscription(LaserScan, 'scan', self.scan_callback)
+
+        self.reset_simulation = node_.create_client(Empty, 'reset_simulation')
+        self.reset_world = node_.create_client(Empty, 'reset_world')
+        self.unpause_proxy = node_.create_client(Empty, 'unpause_physics')
```
Comment (SteveMacenski) [inline nav2_experimental_rl/turtlebot3_env.py]: These proxys arent being used. This is probably important so that the simulation pauses between updates to simulate actual N hz control, the differential drive plugin in gazebo has no timeout entry so the last command will make the robot go in that cmd_vel forever or until a new command is processed. For a mobile robot its less noticable than a manipulator that would drop when it didn't receive commands fast enough but it should be done so that the training looks closer to reality in running on hardware.

I.e. if you want control efforts at 20hz, every time you send a command, let it forward simulate 1/20 of a second, the pause, run through control efforts, unpause, send command, let it forward simulate for 1/20 of a second, and repeat. This will look in simulation time closer to reality even if unintuitive.

File: nav2_experimental_rl/turtlebot3_env.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        get_reward = self.get_reward()
+        return self.states_input, get_reward[0], self.done
+    
+    def check_collision(self):
+        if  min(self.laser_scan_range) < self.range_min + self.collision_tol:
+         print("Near collision detected... " + str(min(self.laser_scan_range)))
+         return True
+        return False
+
+    def reset(self):
```
Comment (SteveMacenski) [inline nav2_experimental_rl/turtlebot3_env.py]: I found in this case, you should wait for measurements before exiting. ie use the `wait_for_message` type primative, that way you dont start processing control efforts before you have new information from your new trial. 0-ing out will probably get rid of issues were stagnated data are being processed in a new run, but it doesn't represent the actual envrionment you're training the agent to react to

**Comment by SteveMacenski** [comment]
@mhpanah I have built today a gazebo plugin for any given robot that will detect collisions using either a simplified collision box via URDF or meshes for ROS1. I've just tested it on my setup so we can stop using laser scanners. Let me know how you'd like me to send it to you if you would like it.

File: nav2_experimental_rl/turtlebot3_env.py (hunk 4 above)
Comment (SteveMacenski) [reply nav2_experimental_rl/turtlebot3_env.py]: Example snippet from my code:
```
        self.unpause_proxy()
        tStart = rospy.Time.now() #ros time so if simulation running at not 1x speed it will be simulated correctly
        self.pub_cmd_vel.publish(vel_cmd)
        while not rospy.is_shutdown():
          should_exit = rospy.Time.now() - tStart > self.control_duration
          if should_exit:
            break
        self.pause_proxy()
```

**Comment by mhpanah** [comment]
> @mhpanah I've also been playing around with this type of tech the last few weeks, good to see similarities in both approaches.
> 
> I wanted to ask if perhaps this doesn't belong in this repo and start a `navigation2_experimental` repo to place this in for the time being. This aligns with out much of the industrial drivers work. I wouldn't feel too comfortable with having this included in the navigation stack directly for the following reasons:
> 
> * assumes a horizontal laser scanner which isn't generalized across all robots (pretty weak reason, but  adding)
> * really only works for the turtlebot3 if using pre-trained weights
> * collision detection based on distance to laser rather than a robust state validity check
> * and the main reason: this technology itself is very experimental and I wouldn't want someone to think this was at a product grade level where they could trust this around people in dynamic environments without severe risk.

@SteveMacenski Yes, I totally agree.  The main purpose of this package is to be used as an example for RL purposes.  We added nav2_experimental directory so that we can put new algorithms or different modules in it for experimental purposes only.  Once we validate and test it on different platforms and collected sufficient data to verify its safety and performance over the existing methods only then we can move it out of the experimental directory. We can either have experimental directory within navigation2 or as you suggested we can have a different repo call it navigation2_experimental.  One benefit of having experimental directory within this repo is to ensure the codes are at lease being reviewed in a timely manner and test it directly within the rest of stack.  @crdelsey @mkhansen-intel @mjeronimo Your thoughts?

**Comment by mhpanah** [comment]
> @mhpanah I have built today a gazebo plugin for any given robot that will detect collisions using either a simplified collision box via URDF or meshes for ROS1. I've just tested it on my setup so we can stop using laser scanners. Let me know how you'd like me to send it to you if you would like it.

@SteveMacenski That's great. Please feel free to either email it to me at mohammad.haghighipanah@intel.com or put instructions here. Thanks.

**Comment by SteveMacenski** [comment]
@mhpanah a `nav2_experimental` directory containing `nav2_dqn` or similar with all the experimental packages would be good as well. I'm pulling the experimental repo structure based on legacy (navigation had a navigation_experimental repo, ros-industrial has a abb_experimental for new drivers, etc) but I prefer your suggestion.

I see you include weights, do those work for non-turtlebot3s as long as generally differential drive? I wouldn't think so since larger robots wouldn't repel obstacles enough since the agent was trained to fail after getting closer than X for turtlebot3, but for a robot 3x the size (like a Tally, Fetch, etc) would just grind. I might add some warnings about how the weights were generated and assumptions with them.

File: nav2_experimental_rl/parameters.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+MEMORY_SIZE = 1000000
+BATCH_SIZE = 64
+EXPLORATION_MIN = 0.25
+EXPLORATION_MAX = 1.0
+EXPLORATION_TARGET_STEP = 1000
+
+# model train parameters
+TARGET_MODEL_UPDATE_STEP = 100 # Number of steps before updating the target model
+EPISODES = 10000 # Desired number of episodes to run
+LOOP_RATE = 0.2 # Control loop rate in seconds
```
Comment (SteveMacenski) [inline nav2_experimental_rl/parameters.py]: loop rate (0.2) or loop duration (5hz)?

File: nav2_experimental_rl/README.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+Firtst run Gazebo:
+```sh 
+export <directory_for_turtlebot3_ws>/src/turtlebot3/turtlebot3_simulations/turtlebot3_gazebo/models
+export TURTLEBOT3_MODEL=burger or waffle
+ros2 launch turtlebot3_gazebo turtlebot3_world.launch.py
+```
+Then, 
+```sh
+cd <navigation2_ws> source install setup.bash
+cd  install/turtlebot3_rl/bin
```
[the remark this answers is not in the document]
Comment (mhpanah) [reply nav2_experimental_rl/README.md]: That won't work since the executable is installed into `/bin`.  However, I should be able to change it so that I can use `ros2 run`.  I will follow the advise from here: https://discourse.ros.org/t/ros2-cannot-run-my-python-script-no-executable-found-error/3912/2

File: nav2_experimental_rl/random_crawl.py (hunk 1 above)
Comment (mhpanah) [reply nav2_experimental_rl/random_crawl.py]: It seems like `get_package_share_directory` is  ROS2 equivalent of rospack.  https://answers.ros.org/question/288501/ros2-equivalent-of-rospackagegetpath/

File: nav2_experimental_rl/turtlebot3_env.py (hunk 4 above)
Comment (mhpanah) [reply nav2_experimental_rl/turtlebot3_env.py]: @SteveMacenski Actually I put pause/unpause proxies there for this intention and also being able to run at x times real-time. But at the time that I implemented this code, I couldn't get current time from rclpy.  Maybe now time.now() is available in rclpy.  Let me check and if it is I will update the code.  Thanks.

Comment (SteveMacenski) [reply nav2_experimental_rl/turtlebot3_env.py]: OK. Another piece of advise I can give you from having done this is that you need to make sure you unpause it on exit conditions so that new sensor readings are able to get received on the first iteration of the new session. Since we send a stop command on collision/goal we're not at risk of getting erroneous simulation artifacts.

File: nav2_experimental_rl/turtlebot3_env.py (hunk 5 above)
Comment (mhpanah) [reply nav2_experimental_rl/turtlebot3_env.py]: It seems like in rclpy 'wait_for_message' is not available.  I used a boolean flag in the subscriber callback as a temporary workaround.

File: nav2_experimental_rl/random_crawl_train.py (hunk 3 above)
Comment (mhpanah) [reply nav2_experimental_rl/random_crawl_train.py]: Fixed.

File: nav2_experimental_rl/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhpanah) [reply nav2_experimental_rl/README.md]: Fixed.

File: nav2_experimental_rl/package.xml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  <name>turtlebot3_rl</name>
+  <version>0.0.1</version>
+  <description>This package enables Reinfocement Learning with Gazebo and Turtlebot3</description>
+
+  <maintainer email="mohammad.haghighipanah@intel.com">Mohammad Haghighipanah</maintainer>
+  <license>Apache License 2.0</license>
+
+  <exec_depend>rclpy</exec_depend>
+  <exec_depend>std_msgs</exec_depend>
+  <exec_depend>keras</exec_depend>
```
[the remark this answers is not in the document]
Comment (mhpanah) [reply nav2_experimental_rl/package.xml]: Fixed.

File: nav2_experimental_rl/random_crawl.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    path = os.path.join(\
+        project_path,"../../nav2_experimental_rl/saved_models/random_crawl_waffle.h5")
+    model = load_model(path)
+    while True:
```
[the remark this answers is not in the document]
Comment (mhpanah) [reply nav2_experimental_rl/random_crawl.py]: Fixed.

File: nav2_experimental_rl/random_crawl.py (hunk 1 above)
Comment (mhpanah) [reply nav2_experimental_rl/random_crawl.py]: Fixed.

File: nav2_experimental_rl/random_crawl_train.py (hunk 2 above)
Comment (mhpanah) [reply nav2_experimental_rl/random_crawl_train.py]: Fixed.

File: nav2_experimental_rl/turtlebot3_env.py [hunk 10]
```diff
@@ -138,11 +140,12 @@ def reset(self):
         while not self.reset_simulation.wait_for_service(timeout_sec=1.0):
             print('Reset simulation service is not available...')
         self.reset_simulation.call_async(Empty.Request())
-        sleep(2)
+
         self.laser_scan_range = [0] * 360
         self.states_input = [3.5] * 8
+        while not self.scan_msg_received and rclpy.ok():
```
Comment (SteveMacenski) [inline nav2_experimental_rl/turtlebot3_env.py]: while true looks need some throttling or this is going to consume a core while waiting

Comment (mhpanah) [reply nav2_experimental_rl/turtlebot3_env.py]: Fixed.

File: nav2_experimental_rl/turtlebot3_env.py (hunk 4 above)
Comment (mhpanah) [reply nav2_experimental_rl/turtlebot3_env.py]: Will do this in another PR once the rclpy have the missing functionality.

**Comment by mhpanah** [comment]
@SteveMacenski I addressed most of your concerns. Could you please take another look. Thanks.

File: nav2_experimental_rl/random_crawl.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+from time import sleep
+from keras.models import load_model
+
+def loadModel(env):
+    state = env.reset()
+    observation_space = len(state)
+    state = np.reshape(state, [1, observation_space])
+    pkg_share_directory = get_package_share_directory('turtlebot3_rl')
+    path = os.path.join(\
+        pkg_share_directory,"../../../../nav2_experimental_rl/saved_models/random_crawl_waffle.h5")
```
Comment (SteveMacenski) [inline nav2_experimental_rl/random_crawl.py]: this seems unideal, it looks like you're backing out of the turtlebot3 directory into nav2_experimental, why not just start there as a base?

File: nav2_experimental_rl/random_crawl.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    model = load_model(path)
+    while rclpy.ok():
+        q_values = model.predict(state)
+        action = np.argmax(q_values)
+        next_state, reward, terminal = env.step(action)
+        next_state = np.reshape(next_state, [1, observation_space])
+        state = next_state
+        sleep(parameters.LOOP_RATE)
```
Comment (SteveMacenski) [inline nav2_experimental_rl/random_crawl.py]: Nothing we can do about this right now, but this should be using a rate object when that's in rclpy

File: nav2_experimental_rl/README.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
+# Overview
+The ROS 2 Experimental Reinfocement Learning (RL) package is a package that enables to train and control a robot with RL techniques.  This package is utilizing Gazebo simulator and ROS 2 to train TurtleBot 3 robot.
```
Comment (mjeronimo) [inline nav2_experimental_rl/README.md]: "This package is utilizing **the** Gazebo simulator and ROS 2 to trans **a** TurtleBot 3 robot."

File: nav2_experimental_rl/random_crawl.py (hunk 11 above)
Comment (mhpanah) [reply nav2_experimental_rl/random_crawl.py]: Fixed.
